File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;

class Solution {
    static class Node {
        String str;
        int cnt;
        
        Node(String str, int cnt) {
            this.str = str;
            this.cnt = cnt;
        }
    }
    
    public int solution(String begin, String target, String[] words) {
        int answer = 0;
        
        answer = bfs(begin, target, words);
        
        return answer;
    }
    
    static int bfs(String begin, String target, String[] words) {
        Queue<Node> q = new LinkedList<>();
        q.offer(new Node(begin, 0));
        
        ArrayList<String> visited = new ArrayList<>();
        
        while(q.isEmpty() == false) {
            Node node = q.poll();
            
            // 정답 반환
            if(node.str.equals(target)) {
                return node.cnt;
            }
            
            for(String word: words) {
                // 방문한적 있는 단어 skip
                if(visited.contains(word)) continue;
                
                int cnt = 0;
                
                for(int i = 0; i < node.str.length(); i++) {
                    if(node.str.charAt(i) != word.charAt(i)) {
                        cnt++;
                    }
                }
                
                if(cnt == 1) {
                    visited.add(word);
                    q.offer(new Node(word, node.cnt+1));
                }
            }
        }
        
        return 0;
    }
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.js
function solution(begin, target, words) {
    var answer = 100;
    
    const dfs = (lev = 0, cur_word = begin, vis = []) => {
        if(lev >= words.length) {
            if(cur_word === target) {
                answer = Math.min(answer, lev)
            }
            return
        }
        if(cur_word === target) {
            answer = Math.min(answer, lev)
            return
        }
        
        for(let i = 0; i < words.length; i++) {
            let cnt = 0
            for(let j = 0; j < cur_word.length; j++) {
                if(cur_word[j] !== words[i][j]) cnt++
            }
            if(cnt !== 1) continue
            if(vis.includes(words[i])) continue
            // console.log(cur_word, "->", words[i])
            vis.push(words[i])
            dfs(lev+1, words[i], vis.slice())
        }
    }
    
    dfs()
    
    answer  = answer === 100 ? 0 : answer
    
    return answer;
}